### PR TITLE
`azurerm_cdn_frontdoor` - Add CDN FrontDoor example configurations

### LINKED ISSUE #10716: Support for CDN Web Application Firewall policies
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hello everyone! 

Azure provides the ability to create WAF policies for CDN Endpoints. This feature is currently in preview status. (https://docs.microsoft.com/de-de/azure/web-application-firewall/cdn/cdn-overview).

Currently, I have to create the policies via the [`az cdn waf policy`](https://docs.microsoft.com/en-us/cli/azure/cdn/waf/policy?view=azure-cli-latest) command afterwards I created the CDN profile + endpoint via terraform. 
I would love to manage this resource type with terraform too!

### New or Affected Resource(s)
This resource type could be a completely new resource type or it can be added to the azurerm_web_application_firewall_policy by adding type for  different types`Azure Front Door` `Azure Appliaction Gateway` & `Azure CDN Endpoints`

new:
* azurerm_cdn_web_application_firewall_policy

 or to the existing:
* azurerm_web_application_firewall_policy

### Potential Terraform Configuration

Extend the existing `azurerm_web_application_firewall_policy` resource with the type of resource the policy is for:
```hcl

resource "azurerm_web_application_firewall_policy" "example" {
...
policy_resource_type: "CDN" // CDN; FrontDoor; ApplicationGateway
...
}
```

### References

* https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/web_application_firewall_policy
* https://docs.microsoft.com/en-us/cli/azure/cdn/waf/policy?view=azure-cli-latest
* https://docs.microsoft.com/de-de/azure/web-application-firewall/cdn/cdn-overview

### LINKED ISSUE #11983: Support for Azure Front Door Standard/Premium
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Would it be possible to add support for [Azure Front Door Standard/Premium](https://docs.microsoft.com/en-us/azure/frontdoor/standard-premium/)?

I manually deployed this resource in the Azure Portal to see what the underlying components are. Unlike the original Front Door resource which is deployed as type "Microsoft.Network/frontdoors", the following resource types are deployed for Front Door Premium:

- Microsoft.Cdn/Profiles
- Microsoft.Cdn/Profiles/AfdEndpoints
- Microsoft.Cdn/Profiles/AfdEndpoints/Routes
- Microsoft.Cdn/Profiles/OriginGroups
- Microsoft.Cdn/Profiles/OriginGroups/Origins

Terraform already supports [azurerm_cdn_profile](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/cdn_profile). However, this Front Door Premium resource is deployed with a SKU of `Premium_AzureFrontDoor`, which is not allowed currently in the "azurerm_cdn_profile" resource:

`Error: expected sku to be one of [Standard_Akamai Standard_ChinaCdn Standard_Verizon Standard_Microsoft Premium_Verizon], got Premium_AzureFrontDoor`

### New or Affected Resource(s)

* azurerm_cdn_profile

### Potential Terraform Configuration

```hcl
resource "azurerm_cdn_profile" "example" {
  name                = "exampleFrontDoorPremium"
  location            = "West US"
  resource_group_name = azurerm_resource_group.example.name
  sku                 = "Premium_AzureFrontDoor"

  origin_group {
    # Origin group config here
  }

  afd_endpoint {
    # Azure Front Door Endpoint config here
  }
}
```

### References

* https://docs.microsoft.com/en-us/rest/api/cdn/cdn/profiles/create

### LINKED ISSUE #13836: CDN FrontDoor: support for Firewall Policies
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The new Azure Frontdoor premium #11983 also uses a different waf SKU. The rest of the waf config seems to be same.
There are now 3 sku's:
- Classic_AzureFrontDoor (default)
- Premium_AzureFrontDoor
- Standard_AzureFrontDoor

### New or Affected Resource(s)

- azurerm_frontdoor_firewall_policy

### Potential Terraform Configuration

```hcl
resource "azurerm_frontdoor_firewall_policy" "example" {
  name                = "example"
  resource_group_name = azurerm_resource_group.rg.name
  enabled                           = true
  mode                              = "Prevention"
  sku                   = "Premium_AzureFrontDoor"
}
```

### References

https://docs.microsoft.com/en-us/rest/api/frontdoorservice/webapplicationfirewall/policies/create-or-update#sku

### LINKED ISSUE #15205: Support for TF resource AFD Standard/Premium availability
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Our partners will need to deploy AFD Standard/Premium with Terraform and they are already asking about the resource availability  

### New or Affected Resource(s)

* azurerm_frontdoor

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

### LINKED ISSUE #16344: Support for Azure Front Door Standard/Premium SKU
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The new Azure Front Door Standard/Premium solution reached GA at the end of March 2022.
https://azure.microsoft.com/en-gb/updates/generally-available-the-new-azure-front-door-a-modern-cloud-cdn-service/

This offers a range of new functionality that the 'Classic' version does not. 

Is it possible to get modules to support deployment and management of Front Door Standard and Premium SKU's via Terraform?

I can see this is linked in https://github.com/hashicorp/terraform-provider-azurerm/pull/15140 and https://github.com/hashicorp/terraform-provider-azurerm/issues/11983 but no current eta of arrival / similar.

Thanks 

### New or Affected Resource(s)/Data Source(s)

azurerm_cdn_frontdoor_*

### Potential Terraform Configuration

_No response_

### References

https://azure.microsoft.com/en-gb/updates/generally-available-the-new-azure-front-door-a-modern-cloud-cdn-service/

### LINKED ISSUE #17460: Support for azurerm_frontdoor
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hello guys,

Is it possible to add the "Pricing Tier" argument into the azurerm_frontdoor resource? 

It looks like it's been added into the AzureRM provider - https://github.com/hashicorp/terraform-provider-azurerm/pull/15140

Thanks guys,
Matt

### New or Affected Resource(s)/Data Source(s)

azurerm_frontdoor

### Potential Terraform Configuration

```hcl
pricing_tier = ["Standard", "Premium", "Classic"]
```

### References

_No response_

### LINKED ISSUE #17721: support for Azure Frontdoor standard/premium
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently the azurerm terraform provider has a azurerm_frontdoor resource. This resource however creates a Azure Frontdoor instance in the 'Classic' SKU. It is currently not possible to create an Azure Front Door in the new Standard/Premium sku.

### New or Affected Resource(s)/Data Source(s)

azurerm_frontdoor

### Potential Terraform Configuration

```hcl
Property and configuration-wise this should be close to the [bicep config](https://docs.microsoft.com/en-us/azure/frontdoor/create-front-door-bicep?tabs=CLI#review-the-bicep-file)
```

### References

_No response_

### PR DESCRIPTION
**Pull request by WodansSon**
## New Resources for CDN Frontdoor

| **Resource Name**                                    | **Current Status** | **PR ID** | **PR Status** |
|-----------------------------------------------|---------------------|------------|----------------|
| azurerm_cdn_frontdoor_profile                   | Done | #17061 | Merged |
| azurerm_cdn_frontdoor_origin                    | Done | #17089 | Merged |
| azurerm_cdn_frontdoor_origin_group         | Done | #17089 | Merged |
| azurerm_cdn_frontdoor_custom_domain    | Done | #18231 | Merged |
| azurerm_cdn_frontdoor_endpoint               | Done | #17078 | Merged |
| azurerm_cdn_frontdoor_route                     | Done | #18231 | Merged |
| azurerm_cdn_frontdoor_route_disable_link_to_default_domain | Done | #18231 | Merged |
| azurerm_cdn_frontdoor_rule_set                 | Done | #17094 | Merged |
| azurerm_cdn_frontdoor_rule                       | Done | #18010 | Merged |
| azurerm_cdn_frontdoor_security_policy      | Done | #17715 | Merged |
| azurerm_cdn_frontdoor_secret                    | Done | #18010 | Merged |
| azurerm_cdn_frontdoor_firewall_policy       | Done | #17715 | Merged |

## ~~New Resources for Custom Domain Workflow Management~~ 

| **Resource Name**                                                                   | **Current Status** | **PR ID** | **PR Status** |
|--------------------------------------------------------------------|---------------------|------------|----------------|
| ~~azurerm_cdn_frontdoor_custom_domain_txt_validator~~            | ~~Done~~                     | ~~#18231~~ | ~~In Review~~ |
| ~~azurerm_cdn_frontdoor_custom_domain_secret_validator~~      | ~~Done~~                     | ~~#18231~~ | ~~In Review~~ |

* `azurerm_cdn_frontdoor_route` - I have moved all of the attributes from the `route association` resource back into the parent(e.g. `route`) resource.  I now poll in the `secret validator` for both the TLS Settings (meaning the Frontdoor service has successfully associated the custom domain with the route and that the SSL cert has been staged to the Domain) and the provision state of the Custom Domain SSL cert.

* ~~`azurerm_cdn_frontdoor_custom_domain_txt_validator` - `azurerm_cdn_frontdoor_custom_domain_secret_validator` will have an internal hard dependency on this resource to keep it from checking the secret state before the Domain has been validated and entered the `approved` state.~~

* ~~`azurerm_cdn_frontdoor_custom_domain_secret_validator` - This resource is used to keep the CNAME record from being written to the DNS host before the certificate has been provisioned to the Custom Domain.~~

~~**NOTE:** I also believe that this will fix issue #11231 because this will introduce a hard dependency chain which will force the CNAME record to be deleted first when deleting the custom domain resource. I have not tested this yet, as I am still developing the resources, but in my head this should fix that issue.~~

**Additional Additional:** With regards to if this will "fix issue #11231". Yes and No... there is a caveat here... it will fix issue #11231 if you bring your DNS `TXT` and `CNAME` records into Terraform so Terraform can manage the relationships. I have tested this extensively and it does work and you no longer receive the error if you configure Terraform in this way. If you leave this as a one off manual task, you will still hit issue #11231. While this is not a perfect solution for the issue, it is a step in the right direction and I believe this will solve many, but not all, customer headaches when dealing with the breaking API change that was introduced by this change in behavior from the service team.

## Test Cases for CDN Frontdoor

| **Resource Name**                                                            | **Test Status** |
|---------------------------------------------------------------|---------------------|
| azurerm_cdn_frontdoor_profile                                          | Passing |
| azurerm_cdn_frontdoor_origin                                           | Passing |
| azurerm_cdn_frontdoor_origin_group                                | Passing |
| azurerm_cdn_frontdoor_custom_domain                           | Skip      |
| azurerm_cdn_frontdoor_endpoint                                      | Passing |
| azurerm_cdn_frontdoor_route                                            | Passing |
| azurerm_cdn_frontdoor_route_disable_link_to_default_domain | Passing |
| azurerm_cdn_frontdoor_rule_set                                        | Passing |
| azurerm_cdn_frontdoor_rule                                              | Passing |
| azurerm_cdn_frontdoor_security_policy                             | Passing |
| azurerm_cdn_frontdoor_secret                                           | Skip      |
| azurerm_cdn_frontdoor_firewall_policy                              | Passing |
| ~~azurerm_cdn_frontdoor_custom_domain_txt_validator~~      | ~~Skip~~      |
| ~~azurerm_cdn_frontdoor_custom_domain_secret_validator~~ | ~~Skip~~      |

## Additional PR Information

**NOTE:** Supersedes existing community PR #14639 and #15140 for same resource(s).

**NOTE:** I had to update the embedded Track1 Azure SDK for GO to workaround an issue with the Service Teams implementation of disabling caching. In the API. You must pass an explicit **NULL** value for the cache field in order to disable the caching behavior in Frontdoor, however, the SDK makes this impossible by excluding the possibility of passing a **NULL** value for the field which controls if caching is enabled or not. To workaround this issue I have removed the **NULL** value check from the SDK for that specific field only which now allows the Terraform AFDx resource to disable caching once it has been enabled. Hopefully, the Pandora SDK will expose a way to override this sort of behavior natively in the form of a switch or feature flag so we can override this sort of non-friendly SDK Service team code moving forward.  

I also had to modify the SDK to allow route to pass **NULL** to remove the custom domain associations.

**NOTE:** Due to a bug in the service teams code, if you are attempting to destroy a `Private Link Endpoint` connection to a `azurerm_lb`(e.g. `Azure Load Balancer`) resource within your CDN Frontdoor resource which requires you to standup your own `azurerm_private_link_service`. You **MUST** add a `depends_on` attribute to your `azurerm_cdn_frontdoor_profile` which references the `azurerm_private_link_service`(e.g. `depends_on = [azurerm_private_link_service.example]`). If this step is skipped you will receive a `Private link service /subscriptions/{subscription}/resourceGroups/{resourceGroup}/providers/Microsoft.Network/privateLinkServices/{privateLinkService} cannot be deleted since it has N private endpoint connections. Please delete all private endpoint connections before deleting the private link service.` error.

---

(fixes #15205 )
(fixes #13836 )
(fixes #11983 )
(fixes #16344)
(fixes #17460)
(fixes #17721)
(fixes #10716)

### CODE REVIEW

**Comment by WodansSon** [comment]
>hey @WodansSon
>
>Apologies for the delayed review here, unfortunately this PR is large enough that Github locks up when attempting to review it - so I've had to resort to pulling this down locally and adding TODO's as review comments.
>
>In short, this PR's off to a good start, but there's a number of changes needed to get this in, including:
>
>We can remove (most of) the embedded SDK's and instead use the Azure SDK with a handful of overrides as required (the pattern we use is to duplicate the specific SDK method in question in the azuresdkhacks package, overriding the Preparer/Responder as required).
>Some of the fields within the resources aren't returned by the API/aren't used - so we'll need to fix/remove those.
Discriminator expand functions - we can use the Azure SDK return types directly (e.g. *SomeType is an implementation of the parent interface SomeInterface [note: no pointer]) so we can return / use the SomeType struct directly.
>
>The Create and Update methods want splitting - where the separate Update method exists we'll also want to use conditional/patch updates (I've pushed a couple of these) using d.HasChange("field") to only update fields which have changed.
>
>All properties within nested lists/sets need to be set into the state.
>
>resourceCdnFrontdoorProfile
resource_group_name can be removed
sku name should be required
cdn_frontdoor_id in read func
response_timeout_seconds can be updated but is ForceNew?
resourceCdnFrontdoorSecret - presumably customer_certificate should be MaxItems: 1?
We should remove all of the provisioning_state fields
Resource timeouts / tests covering custom domain
resourceCdnFrontdoorFirewallPolicy
SKU should be required
Minor: moving the validators into validators, key vault ids etc
>
>Whilst I haven't called them out explicitly here, I have a feeling there's also some schema changes required to the Discriminated items, but we'll need to fix the Discriminator expand functions before we can determine the appropriate schema there unfortunately.
>
>Since there's a lot of commits in this branch, I've gone ahead and rebased/squashed these down to make this easier to review - at the same time I've gone ahead and rebased the commits in this branch in order to remove (most of) the Embedded SDK's by making use of the azuresdkhacks workaround above.
>
>For now I've pushed that into r-frontdoor-profile-cleanedup: https://github.com/hashicorp/terraform-provider-azurerm/tree/r-frontdoor-profile-cleanedup.
>
>Note: That branch removes all merge commits in favour of rebasing on top of main, so there's far fewer changes.
>
>After doing that, I've then gone through and fixed some of these issues and left TODO's inline (and calling out several blockers) for the others which we'll need to address.
>
>Whilst I appreciate that's an not ideal way to review this, given Github is locking up this seemed like the best way to review this for now - but if you can work through the TODO's then we can take another look 👍
>
>Thanks!

@tombuildsstuff I am looking over the changes, thanks.

**Comment by sebader** [comment]
@WodansSon I'm still trying to work around the issue that a route would get destroyed and recreated when I'm updating all the origins for that route. Reason being is that this recreation always created a couple of minutes outage for my Front Door - which we of course need to avoid at all costs.

So my idea was to not change the name of the origins (basically just using indexes as a the name) and update all the properties of the origin:

image


The apply for the second origin (staticstorage) works fine. For the "backendapi" it also seems to get applied fine on the Front Door resource. Private Endpoint gets created and all looks good. However the apply just times out with this error:

    │ Error: waiting for the update of Frontdoor Origin: (Origin Name "0" / Origin Group Name "BackendApis" / Profile Name "sebce251d9-global-fd" / Resource Group "sebce251d9-global-rg"): Code="InternalError" Message="Sorry, it looks like there was an error on our end. Please contact Support if you keep having this problem."


image

Any ideas? Or is there a way we can work around the actual problem when replacing all the origins for an origin group? Recreating the route and causing an outage is really not good.

**Comment by Evilweebl** [comment]
I'm getting an error trying to create the azurerm_cdn_frondoor_origin resource. I'm following the example [here](https://github.com/hashicorp/terraform-provider-azurerm/pull/16671/files#diff-b1228c74223045650a2b49e349fd78343eb348c937a22eff7955b61852614d6c) and it's throwing an error saying an underlying private link attribute is required, but I've not defined a private_link block because I don't want it.

`Error: creating Frontdoor Origin: (Origin Name "test-origin" / Origin Group Name "test-origin-group" / Profile Name "test-profile" / Resource Group "test-rg"): cdn.AFDOriginsClient#Create: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="Property 'AfdOrigin.SharedPrivateLinkResource.PrivateLink' is required but it was not set; \nProperty 'AfdOrigin.SharedPrivateLinkResource.RequestMessage' is required but it was not set"`

**Comment by WodansSon** [comment]
@Evilweebl Thank you for your comment. This has been fixed with my `d38c13b` commit. This was caused by returning an empty `sharedPrivateLinkResource` object to the API instead of a explicit `nil` value. The service saw that it wasn't `nil` and attempted to enumerate it which caused the error you saw. With my most recent check in I reverted that code to return `nil` again which will fix that error you saw. Most services interpret empty objects as `nil` however, Frontdoor does not. 🤷‍♂️

**Comment by WodansSon** [comment]
These resources should be pretty close to being shippable at this point, however I need to verify that the disable cache still works since my custom track 1 SDK has been reverted... I might need to add another `azurehacks` file to override the existing Track 1 SDK... short of any requested schema changes that maybe requested I think this is pretty close... 🙂

**Comment by WodansSon** [comment]
@tombuildsstuff  Per your comment: `All properties within nested lists/sets need to be set into the state.` is this a new directive for pandora? I have been working on this provider for 7 years now and that is the exactly opposite of what was previously expected. Why the change?

**Comment by WodansSon** [comment]
@Evilweebl no worries... totally taken as constructive and not nagging at all. 

>Firstly, I believe the docs regarding the [azurerm_cdn_frontdoor_custom_domain ](https://github.com/hashicorp/terraform-provider-azurerm/pull/16671/files?short_path=9ec93fe#diff-dc24090ab19321c52aa09ad0e3b605b116b0166005dbf9241c8366ddeac9070c)are incorrect. After checking my state it appears there is no validation_properties block, The attributes expiration_date and validation_token are part of the flat structure. This might mean changing docs [here ](https://github.com/hashicorp/terraform-provider-azurerm/blob/cb2cb1eac11efda9c61b2b146f042a8d5fd2467f/website/docs/r/cdn_frontdoor_custom_domain_txt_validator.html.markdown?plain=1#L61)as well as other places.

Yep, that was part of the PR review I am working through at the moment. I flattened the structure, but I forgot to update the docs. I will fix that shortly.

>Secondly, I had some questions around the `azurerm_cdn_frontdoor_custom_domain_txt_validator`. In the examples it focuses on use of the azure dns zone but for my use case I am using cloudflare. The dns_txt_record_id argument is a required field and expects an azure resource id which in my case I don't have. Should this be optional or should I just not need to create a txt validator resource. It's worth pointing out that the latter would also cause issues with the secret validator as that requires a text validator id.

I thought of that, but I pushed it off until later and I guess later is here now. 🙂 The reason I exposed the txt validator and the secret validator is because in the new Frontdoor service the service team not has created a workflow where:
* You must prove that you own the domain via a `_dnsauth` txt record in the DNS entries of the apex domain you are attempting to provision a custom domain for.
* Once you have proven you own the domain with the txt record you then need to associate the new custom domain with a Frontdoor route.
* Once you have your route associated to your custom domain the Frontdoor service will then provision the Customer supplied SSL cert or the Frontdoor managed SSL cert to the custom domain. This is where the secret validator comes it, it waits until the Frontdoor service confirms that that SSL cert has been provisioned to the custom domain.
* Lastly, once the SSL cert has been provisioned you can then add your CNAME record which will enable the traffic to flow as expected.

Since with Terraform you are supposed to be able to set up the service in a single `apply` as you can see from the above I had to get a bit creative in order to make that happen. The reason I exposed the `dns_txt_record_id` is to order when Terraform should create the resource and start polling the service see if the custom domain has been `approved` yet or not. Now that I think about it, I prolly don't even need that field since the only thing that will happen is that the poll may get a few more `pending` responses from the service if it creates the txt validator right after the custom domain is created. Which I think is acceptable, and it would work for both dns zone and non-dns zone scenarios.

File: website/docs/r/cdn_frontdoor_security_policy.html.markdown
Comment (sebader) [inline website/docs/r/cdn_frontdoor_security_policy.html.markdown:29]:
```suggestion
```

File: website/docs/r/cdn_frontdoor_firewall_policy.html.markdown
Comment (sebader) [inline website/docs/r/cdn_frontdoor_firewall_policy.html.markdown:204]: can you add what the default value is if this is omitted?

**Review by sebader — CHANGES_REQUESTED** [review]

File: website/docs/r/cdn_frontdoor_security_policy.html.markdown
Comment (WodansSon) [reply website/docs/r/cdn_frontdoor_security_policy.html.markdown:29]: Ahhh... good catch fixed... I thought I got all of those but apparently I missed one... Thanks 🚀

File: website/docs/r/cdn_frontdoor_firewall_policy.html.markdown
Comment (WodansSon) [reply website/docs/r/cdn_frontdoor_firewall_policy.html.markdown:204]: Yep... fixed... BTW it `Block` 🚀

File: website/docs/r/cdn_frontdoor_security_policy.html.markdown
Comment (sebader) [reply website/docs/r/cdn_frontdoor_security_policy.html.markdown:29]: not sure if you meant to take it out yet but it is still in there ;)

**Comment by WodansSon** [comment]
image

File: website/docs/r/cdn_frontdoor_security_policy.html.markdown
Comment (WodansSon) [reply website/docs/r/cdn_frontdoor_security_policy.html.markdown:29]: @sebader lol... I was looking at the firewall policy documentation not the security policy documentation... I got it now... 🤣

File: website/docs/r/cdn_frontdoor_route.html.markdown
Comment (sebader) [inline website/docs/r/cdn_frontdoor_route.html.markdown:132]: should we add a note that this only work if `supported_protocols` has both set, HTTP and HTTPS. Tooke me a while to realize this :D not super intuitive

Comment (WodansSon) [reply website/docs/r/cdn_frontdoor_route.html.markdown:132]: @sebader, good point... I note would be helpful and I should also add a validation rule that verifies that these setting are set correctly and raise an error if the resource will not be doing what you expect it to be doing, because I agree it is not very intuitive to create the route resource correctly.

Comment (sebader) [reply website/docs/r/cdn_frontdoor_route.html.markdown:132]: yep sounds good

**Comment by WodansSon** [comment]
image

**Comment by sebader** [comment]
@tombuildsstuff any update on when we could get this merged? I've also tested this quite extensively in various configurations. With the latest changes @WodansSon did it all worked pretty well for me.

File: website/docs/r/cdn_frontdoor_origin.html.markdown
Comment (sebader) [inline website/docs/r/cdn_frontdoor_origin.html.markdown:143]: as stated in the comment above, currently "web" is not allowed. Not sure if we just want to leave it in anyway in anticipation of this hopefully coming back soon?!

But I think you can remove the TBD. And maybe reword to something like "when using Load Balancer / private link service, do not set this field"

File: website/docs/r/cdn_frontdoor_origin.html.markdown
Comment (sebader) [inline website/docs/r/cdn_frontdoor_origin.html.markdown:145]: should we add a link to the docs which locations are possible https://docs.microsoft.com/en-us/azure/frontdoor/private-link#region-availability

**Review by sebader — CHANGES_REQUESTED** [review]

**Comment by jrjohnson87** [comment]
Hi, is there an ETA/milestone for when the additional CDN Frontdoor resources will be available?

**Comment by LWNetworks** [comment]
Echoing @jrjohnson87 - we're currently waiting for this internally to enable a full IaC deployment. 
@tombuildsstuff do you have a rough ETA for review/approval and release in the latest provider version? Thanks

**Comment by WodansSon** [comment]
@jrjohnson87 and @LWNetworks I cannot commit a hard ETA for these resources to be 100% released due to circumstances that are outside of my control, but it is my personal goal is to have the rest of the resources released with the official provider over the next 4 to 5 release cycles. I was recently given full control of these resources again and it is my top priority. I am sorry for the delay, and I appreciate your patience.

**Comment by p-m-j** [comment]
is the order known for the remaining resources, i'm keen to make use of `azurerm_cdn_frontdoor_route` any chance that's coming next?

**Comment by cstickel** [comment]
@p-m-j a quick hint that can originally be found in the container apps issue. using azapi provider is a nice way to use resources where azurerm provider support is still missing. https://registry.terraform.io/providers/azure/azapi/latest/docs
it's not as comfortable, but works quite fine and a lot better than the frustration to be blocked while waiting.

**Comment by WodansSon** [comment]
@p-m-j 

> is the order known for the remaining resources, i'm keen to make use of `azurerm_cdn_frontdoor_route` any chance that's coming next?

Unfortunately, no `azurerm_cdn_frontdoor_route` will not be in the next PR due to interdependencies within the resources... once the current PR #17715 gets merged the order of the remaining PR's will be:

PR 1:
* `azurerm_cdn_frontdoor_secret`
* `azurerm_cdn_frontdoor_rule`

PR 2:
* `azurerm_cdn_frontdoor_route`
* `azurerm_cdn_frontdoor_custom_domain`
* `azurerm_cdn_frontdoor_custom_domain_txt_validator`
* `azurerm_cdn_frontdoor_custom_domain_secret_validator`

PR 3 will be to add all of the example configurations I have written to the examples folder.

As I stated above, the speed of these PR's and the resources becoming available are "...due to circumstances that are outside of my control..." and I apologize, but I am doing the best that I can.

**Comment by sschu** [comment]
Hi @WodansSon!
Does PR2 containing `azurerm_cdn_frontdoor_route`, `azurerm_cdn_frontdoor_custom_domain`, `azurerm_cdn_frontdoor_custom_domain_txt_validator`, `azurerm_cdn_frontdoor_custom_domain_secret_validator`already exist? I could not find it and it is not linked? Do you have a rough estimate when these resources would be available in the provider? Thanks!

**Comment by WodansSon** [comment]
@sschu 

> Hi @WodansSon! Does PR2 containing `azurerm_cdn_frontdoor_route`, `azurerm_cdn_frontdoor_custom_domain`, `azurerm_cdn_frontdoor_custom_domain_txt_validator`, `azurerm_cdn_frontdoor_custom_domain_secret_validator`already exist? I could not find it and it is not linked? Do you have a rough estimate when these resources would be available in the provider? Thanks!

Yes... I have just opened PR #18231 with these resources since #18010 was just merged for the v3.21.0 provider release. I am now actively working on these last four resources. I do not have an ETA for the availability of these resources, but it is my personal goal to have them available within the next two weeks or so.

**Review by katbyte — APPROVED** [review]
LGTM 🍄

**Pull request merged**
